spec/abstract/Decode_A.thy

```diff
@@ -515,47 +509,62 @@ where
         odE
       | _ \<Rightarrow> throwError (InvalidCapability 1);
       returnOk $ InvokeSchedContextBind sc_ptr cap
-    odE
-  | SchedContextUnbindObject \<Rightarrow> doE
+    odE"
+
+definition
+  decode_sched_context_unbind_object ::
+  "obj_ref \<Rightarrow> cap list \<Rightarrow> (sched_context_invocation, 'z::state_ext) se_monad"
+where
+  "decode_sched_context_unbind_object sc_ptr excaps \<equiv> doE
       whenE (length excaps = 0) $ throwError TruncatedMessage;
       cap \<leftarrow> returnOk $ hd excaps;
       case cap of
         ThreadCap tcb_ptr \<Rightarrow> doE
-          sc_tcb_opt \<leftarrow> liftE $ get_sc_obj_ref sc_tcb sc_ptr;
-          whenE (sc_tcb_opt \<noteq> Some tcb_ptr) $ throwError IllegalOperation;
+          sc \<leftarrow> liftE $ get_sched_context sc_ptr;
+          whenE (sc_tcb sc \<noteq> Some tcb_ptr) $ throwError IllegalOperation;
```

Comment:
I think I just preferred using the `corres` rule for `get_sched_context`. But I'll have another look at these things and give it another shot

Reply:
I think you can just unfold `get_sc_obj_ref` to see `get_sched_context`, maybe with simp or something.

Reply:
@corlewis summoned me here, but it looks like this is already changed back to `get_sc_obj_ref`. :-)

@michaelmcinerney, can you clarify what was preferable about `get_sc_corres`? Is it just that it existed, but `get_sc_ob_ref_corres` did not?

In the long run, I think the optimum would be for every access to be just `get` with a lens (and preferably the most focussed lens possible). Then you only need `corres_get`. For now, we at least have that `get_sc_obj_ref` is better than (hypothetical) `get_sc_tcb`, since the former needs only `get_sc_obj_ref_corres` whereas the latter would need `get_sc_tcb_corres`, `get_sc_yield_to_corres`, etc. And along the lines of @corlewis' comment, `get_sc_obj_ref sc_tcb` is usually preferable to `get_sched_context`, because the latter sometimes means you end up explaining things about fields you didn't actually look at.

And @mktnk3 is also right that in proof contexts where `get_sched_context` does not force you to explain things about other fields, you can just unfold, leaving the spec in the more specific form.

Reply:
It was just that at the time, I knew off the top of my head that there was `get_sc_corres`, but I couldn't remember a similar one for getting a particular field in the sc. And seeing that it was used elsewhere and lined up nicely with the Haskell, I just went for it. I've gone with unfolding `get_sc_obj_ref` and using another rule to rewrite it so that I can then use `get_sc_corres`

Reply:
Right, makes sense.